lib/phoenix/template.ex

```diff
@@ -54,6 +54,130 @@ defmodule Phoenix.Template do
+  end
+
+  @doc """
+  Renders template from module.
+
+  For a module called `MyApp.FooHTML` and template "index.html.heex",
+  it will:
+
+    * First attempt to call `MyApp.FooHTML.index(assigns)`
+
+    * Then fallback to `MyApp.FooHTML.render("index.html", assigns)`
+
+    * Raise otherwise
+
+  It expects the HTML module, the template as a string or atom, the format, and a
+  set of assigns.
+
+  Notice that this function returns the inner representation of a
+  template. If you want the encoded template as a result, use
+  `render_to_iodata/4` instead.
+
+  ## Examples
+
+      Phoenix.Template.render(YourApp.UserView, :index, "html", name: "John Doe")
+      #=> {:safe, "Hello John Doe"}
+
+      Phoenix.Template.render(YourApp.UserView, "index", "html", name: "John Doe")
+      #=> {:safe, "Hello John Doe"}
+
+  ## Assigns
+
+  Assigns are meant to be user data that will be available in templates.
+  However, there are keys under assigns that are specially handled by
+  Phoenix, they are:
+
+    * `:layout` - tells Phoenix to wrap the rendered result in the
+      given layout. See next section
+
+  ## Layouts
+
+  Templates can be rendered within other templates using the `:layout`
+  option. `:layout` accepts a tuple of the form
+  `{LayoutModule, "template.extension"} | {LayoutModule, :template}`.
+
+  To template that goes inside the layout will be placed in the `@inner_content`
+  assign:
+
+      <%= @inner_content %>
+
+  """
+  def render(module, template, format, assigns) do
+    assigns
+    |> Map.new()
+    |> Map.pop(:layout, false)
+    |> render_within_layout(module, template, format)
+  end
+
+  defp render_within_layout({false, assigns}, module, template, format) do
+    render_with_fallback(module, template, format, assigns)
+  end
+
+  defp render_within_layout({{layout_mod, layout_tpl}, assigns}, module, template, format)
+       when is_atom(layout_mod) and (is_binary(layout_tpl) or is_atom(layout_tpl)) do
+    content = render_with_fallback(module, template, format, assigns)
+    assigns = Map.put(assigns, :inner_content, content)
+    render_with_fallback(layout_mod, layout_tpl, format, assigns)
+  end
+
+  defp render_within_layout({layout, _assigns}, _module, _template, _format) do
+    raise ArgumentError, """
+    invalid value for reserved key :layout in Phoenix.Template.render/4 assigns
+    :layout accepts a tuple of the form {LayoutModule, "template.extension"}
+    got: #{inspect(layout)}
+    """
+  end
+
+  defp template_with_ext(template, format), do: to_string(template) <> "." <> format
+
+  defp encode(content, template_with_ext) do
+    if encoder = format_encoder(template_with_ext) do
+      encoder.encode_to_iodata!(content)
+    else
+      content
+    end
+  end
+
+  defp render_with_fallback(module, template, format, assigns)
+       when is_atom(module) and is_atom(template) and is_binary(format) and is_map(assigns) do
+    if function_exported?(module, template, 1) do
```

Comment:
We need the erlang module_loaded check both before String.to_existing_atom and the function_exported? check, so we should probably move it somewhere so we can check it only once:

```suggestion
    :erlang.module_loaded(module) or :code.ensure_loaded(module)

    if function_exported?(module, template, 1) do
```